File: 设计模式/设计模式/命令模式/单品/动态调用/DynamicCommand.m
//
//  DynamicCommand.m
//  设计模式
//
//  Created by zachary spark on 2019/5/13.
//  Copyright © 2019 3g. All rights reserved.
//

#import "DynamicCommand.h"

@interface DynamicCommand()
@property(nonatomic,strong)DanPinRealization *Danpin;
@property(nonatomic,copy)DynamicBlock DanPinBlock;

@end

@implementation DynamicCommand


- (instancetype)init:(DanPinRealization *)Danpin withObject:(id)data block:(DynamicBlock)DanPinBlock {
    
    self = [super init];
    if(self){
        self.data = data;
        self.Danpin = Danpin;
        self.DanPinBlock = DanPinBlock;
    }
    return self;
    
    
}

- (void)execute{
    self.DanPinBlock(self.Danpin,self.data);
}

//创建对象的时候 由于有时候初始化参数的过于复杂，所以我们可以内部提供
//动态命令创建过程，专门有了实现，外部只需要调用即可
+(id<DanPinCommandProtocol>)CreateCommand:(DanPinRealization *)DanPin withObject:(id)data block:(DynamicBlock)DanPinBlock {
    
    return [[DynamicCommand alloc]init:DanPin withObject:data  block:DanPinBlock];
    
}


@end

File: 设计模式-Swift/设计模式-Swift/命令模式/调用添加命令/DynamicCommand.swift
//
//  DynamicCommand.swift
//  设计模式-Swift
//
//  Created by zachary spark on 2019/5/14.
//  Copyright © 2019 3g. All rights reserved.
//

import UIKit

class DynamicCommand: DanPinCommandProtocol {

    private var receiver:DanPinRealization
    
    private var block:(DanPinRealization,AnyObject?)->Void
    
    //由于swift是一门严谨的语言
    //这个动态命令的参数，如果为空，当命令execute的时候就不进行传递参数
    var data:AnyObject?
    
    //定义个构造方法
    init(_ receiver:DanPinRealization ,_ data:AnyObject?,_ block:@escaping (DanPinRealization,AnyObject?)->Void) {
        self.data = data
        self.receiver = receiver
        self.block = block
    }
    
    func execute() {
        self.block(self.receiver,self.data)
    }
    
    //创建动态命令的方法
    class func createCommand(_ receiver:DanPinRealization,_ data:AnyObject?, _ block:@escaping(DanPinRealization,AnyObject?)->Void)->DanPinCommandProtocol{
        
        return DynamicCommand.init(receiver,data,block)
        
    }
    
}
